File: base/transportation.py
import csv
import traceback

from django.contrib.auth.decorators import permission_required
from django.http import HttpResponseBadRequest, JsonResponse, HttpResponse
from django.shortcuts import render

from base.models import UserProfile
from utils.celery import error_log


@permission_required('base.transportation_page')
def home(request):
    return render(request, 'transportation.html')


@permission_required('base.transportation_page')
def get_employees(request):
    try:
        values = ('empid', 'full_name', 'team__shift__intime', 'department__name', 'designation__name', 'gender',
                  'phone', 'current_address', 'permanent_address', 'emergency_contact', 'emergency_phone')
        resp = list(UserProfile.objects.filter(user__is_active=True, transport_required='yes').values(*values).order_by(
            'team__shift__intime'))

        return JsonResponse(resp, safe=False)

    except Exception as error:
        error_log.delay(request.path, request.META['REMOTE_ADDR'], str(traceback.format_exc()))
        return HttpResponseBadRequest(error)


@permission_required('base.transportation_page')
def download_as_csv(request):
    try:
        response = HttpResponse(content_type='text/csv')
        response[
            'Content-Disposition'] = 'attachment; filename="Transportation_List.csv"'

        writer = csv.writer(response)
        writer.writerow(['EMPID', 'Name', 'Shift Time', 'Department', 'Gender', 'Phone', 'Current Address',
                         'Permanent Address', 'Emergency Contact', 'Emergency Phone'])

        values = ('empid', 'full_name', 'team__shift__intime', 'department__name', 'gender',
                  'phone', 'current_address', 'permanent_address', 'emergency_contact', 'emergency_phone')

        employees = list(UserProfile.objects.filter(
            user__is_active=True, transport_required='yes').values_list(*values).order_by('team__shift__intime'))

        for emp in employees:
            writer.writerow(emp)

        return response

    except Exception as error:
        error_log.delay(request.path, request.META['REMOTE_ADDR'], str(traceback.format_exc()))
        return HttpResponseBadRequest(error)

File: static/js/transportation.js
class Employee {
    constructor(data) {
        this.data = data;
        this.element = createElements(el('tr', {},
            el('td', {}, this.data.empid),
            el('td', {}, this.data.full_name),
            el('td', {}, this.data.team__shift__intime),
            el('td', {}, this.data.department__name),
            el('td', {}, this.data.gender),
            el('td', {}, this.data.phone),
            el('td', {}, this.data.current_address),
            el('td', {}, this.data.permanent_address),
            el('td', {}, this.data.emergency_contact),
            el('td', {}, this.data.emergency_phone),
        ))
    }
}

const get_employees = () => {
    let employees_list = document.getElementById('employees_list');

    return post('get_employees/', {}).then(resp => {
        resp.forEach(data => {
            let employee = new Employee(data);
            employees_list.append(employee.element);
        });

        document.getElementById('total_employees').innerText = resp.length;
    })
};

window.onload = async () => {
    show_loading();
    await init();

    get_employees();

    // set title
    document.title = 'Transportation';
    hide_loading()
};